Port over upwind flux for curved scalar waves as a BoundaryCorrection

prayush opened the pull request:
## Proposed changes

This PR ports over upwind penalty flux for the curved scalar wave evolution system to the `BoundaryCorrection` framework. In the process, it modifies the time derivatives' calculator action struct to furnish additional quantities needed by the upwind flux.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

prayush committed:
Modify CSW::time derivs to return vars needed by bndry corrs

nikwit commented:
Have you tried running this with a test executable? I havent looked at the code too much but when I run this with the constraint preserving boundary conditions I ported on a simple sine example with Minkowski background, the whole wave explodes almost immediately everywhere on the domain. When I just copy over the UpwindPenalty from the (flat) `ScalarWave` system it works fine. This is just a very naive substitution so maybe I am doing something wrong though.

Maybe there could be a unit test which ensures that the flux reduces to the flat system for Minkowski?

prayush commented:
> Have you tried running this with a test executable? I havent looked at the code too much but when I run this with the constraint preserving boundary conditions I ported on a simple sine example with Minkowski background, the whole wave explodes almost immediately everywhere on the domain. When I just copy over the UpwindPenalty from the (flat) `ScalarWave` system it works fine. This is just a very naive substitution so maybe I am doing something wrong though.
> 
> Maybe there could be a unit test which ensures that the flux reduces to the flat system for Minkowski?

Hi @nikwit - I found and fixed a major bug in the code here, the internal and external contributions were being added instead of subtracted. I wrote a test comparing this flux class with the flat `ScalarWave`'s upwind flux in flat space, and that helped me discover this bug. After fixing it, that test now passes. This code is now ready for review.

prayush commented:
> > Have you tried running this with a test executable? I havent looked at the code too much but when I run this with the constraint preserving boundary conditions I ported on a simple sine example with Minkowski background, the whole wave explodes almost immediately everywhere on the domain. When I just copy over the UpwindPenalty from the (flat) `ScalarWave` system it works fine. This is just a very naive substitution so maybe I am doing something wrong though.
> > Maybe there could be a unit test which ensures that the flux reduces to the flat system for Minkowski?
> 
> Hi @nikwit - I found and fixed a major bug in the code here, the internal and external contributions were being added instead of subtracted. I wrote a test comparing this flux class with the flat `ScalarWave`'s upwind flux in flat space, and that helped me discover this bug. After fixing it, that test now passes. This code is now ready for review.

I also updated the executable in #2973  to use this upwind flux. We get stable evolutions of plane waves there now. Can I request you to have a quick look at this PR so we can get this in quick please?

nilsvu commented:
@prayush would you mind deleting the old code you are porting over, if it's not needed anymore?

prayush committed:
Remove redundant NumericalFlux UpwindPenaltyCorrection for CurvedScalarWave

prayush commented:
> @prayush would you mind deleting the old code you are porting over, if it's not needed anymore?

@nilsleiffischer : Removed the redundant code too, you're right we don't really need it it anywhere.

nilsdeppe commented on src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp:
```diff
@@ -0,0 +1,231 @@
+
+  {  // package characteristic fields
+    const auto char_fields =
+        characteristic_fields(constraint_gamma2, inverse_spatial_metric, psi,
```
is there a `not_null` version of this function so you can avoid memory allocations? Anything that does memory allocations on the interfaces is rather expensive. The `char_fields` function also does 2 more memory allocations, so overall this part of the boundary correction is already at 4 (including the shift dot normal below) memory allocations per interface plus a bunch of copy of data. It's very little work compared to memory movement.

(see comment below for more thoughts :) )

prayush replied:
using return by reference of this function, AND removed the memory allocation in the `char_fields` function.

nilsdeppe commented on src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp:
```diff
@@ -0,0 +1,231 @@
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal_ext,
+    const tnsr::a<DataVector, 3, Frame::Inertial>& char_speeds_ext,
+    dg::Formulation /*dg_formulation*/) const noexcept {
+  Variables<CurvedScalarWave_detail::char_field_tags<Dim>>
```
this function is also pretty brutal on memory allocations. In general, we should do at most 1 memory allocation per interface in `dg_boundary_terms` and one in `dg_package_data`. 

I'll let you and @nikwit coordinate what you think is best here since he's the one using the evolution system. If the conclusion is "let's merge this PR with all the extra allocations and copies" then it would be good to file an issue so that we remember this is a problem

prayush replied:
Okay, reduced the total number of allocations to 1 buffer allocation. Also reduced the total size of allocation by reusing memory.

nilsdeppe commented on tests/Unit/Evolution/Systems/CurvedScalarWave/CMakeLists.txt:
```diff
@@ -15,5 +16,5 @@ add_test_library(
   "Evolution/Systems/CurvedScalarWave/"
   "${LIBRARY_SOURCES}"
-  "CurvedScalarWave;CurvedScalarWaveHelpers;GeneralRelativityHelpers"
+  "DataStructures;CurvedScalarWave;CurvedScalarWaveHelpers;GeneralRelativityHelpers;Utilities"
```
[optional] You can also change this to
```
  ""
  )

target_link_libraries(
  ${LIBRARY}
  PRIVATE
  # all the libs to link
  )
```
this is easier to read and what we're hoping to switch to in the long term :)

prayush replied:
changed

nilsdeppe reviewed (dismissed):
Looks good, some thoughts that I think you and @nikwit should decide on the best course of action for. If you do end up changing code, please feel free to squash things directly :)

nikwit commented on src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  {  // package characteristic speeds
+    Scalar<DataVector> shift_dot_normal{};
```
There is a function `CurvedScalarWave::characteristic_speeds` which for some reason returns a `std::array<DataVector, 4>`. I would change it to the implementation you have right here and return a tensor, this could then also be called from the boundary conditions and makes `dg_package_data` very concise.

prayush replied:
I would be happy to make this change but I worry this might not play well with other parts of the code, as all evolution systems at the moment use `std::array`'s. A quick grep tells me char speeds are directly used in, for e.g., the fluxes:

```
NumericalAlgorithms/DiscontinuousGalerkin/NumericalFluxes/LocalLaxFriedrichs.hpp
NumericalAlgorithms/DiscontinuousGalerkin/NumericalFluxes/Hll.hpp
```

@nilsdeppe can you please confirm that swapping out the type of characteristic speeds from `std::array<DataVector, 4> --> tnsr::a<3>` not break anything else?

nilsdeppe replied:
Those fluxes aren't used anymore by the evolution code (I don't think by the elliptic code either). However, since every other system's `characteristic_speeds` function uses `std::array`s it seems it would be better not to change it in just one system. It is possible to assign a `std::array` to a `tnsr::a<3>`, but it might have to be done explicitly: `*packaged_char_speeds = tnsr::a<3, DataVector, Frame::Inertial>{characteristic_speeds(...)};`

prayush replied:
I agree that this should change simultaneously in all evolution systems. And that is why I'd rather not call the `CurvedScalarWave::characteristic_speeds` function in `dg_package_data` as that would need an additional allocation of a `std::array<DV, 4>`. Its true this leads to code duplication, but I can't think of a better solution given the constraints.

nikwit replied:
ok, I would argue that avoiding this code duplication is more important than keeping a completely unused function consistent with other unused functions, which are now essentially deprecated. Keep in mind that I will have to duplicate it again for the boundary conditions as will pretty much any other boundary condition / boundary correction implementation.

But if you decide to leave it, I can port this over for all systems before I do the boundary conditions - I dont want to hold this PR up.

prayush replied:
Okay, I am convinced. As a middle ground, I have added an overload of `characteristic_speeds` that returns a `tnsr::a<3>` by reference instead of a `std::array`. I have not deleted the version(s) that return(s) `array`s as that should be done in a different PR for all evolution systems simultaneously. 

This overload solves our purpose in both the flux and boundary conditions for this system.

nikwit replied:
I think that is a good solution

nikwit commented on src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,178 @@
+class UpwindPenalty final : public BoundaryCorrection<Dim> {
+ private:
+  struct CharSpeedsTensor : db::SimpleTag {
+    using type = tnsr::a<DataVector, 3, Frame::Inertial>;
```
`tnsr::a<3>` or `tnsr::i<4>`? Shouldnt make a difference as it is just used for packaging but I would personally prefer the latter.

nilsdeppe replied:
This should be consistent with the other systems, which I think use a `tnsr::a<3>`

nikwit commented on src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp:
```diff
@@ -0,0 +1,231 @@
+        (-sign * step_function(sign * char_speed) * char_speed);
+  }
+
+  return weighted_char_field;
```
you are relying on copy elision here but maybe a return by reference might be safer to avoid a huge amount of allocations if it fails.

prayush replied:
fair point, replaced with return by reference now

nikwit commented on src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/BoundaryCorrection.hpp:
```diff
@@ -0,0 +1,44 @@
+// Distributed under the MIT License.
```
(random line) I think you need to add the tag `using inverse_spatial_metric_tag = gr::Tags::InverseSpatialMetric<volume_dim, Frame::Inertial, DataVector>;` to `System.hpp`, otherwise your normal vectors will be normalized w.r.t. the Euclidean metric resulting in the wrong computation of characteristic speeds, see `PackageDataImpl.hpp:48`. This will then also automatically slice in the normal vector in addition to the covector for reasons I dont understand. I suppose you might then not need the `inverse_spatial_metric` if you change the `characteristic_fields` function to accept the unit vector directly. Maybe @nilsdeppe can explain?

nilsdeppe replied:
if a system has a non-flat Cartesian metric then the `inverse_spatial_metric` tag must be specified in the system, otherwise you'll get incorrect normal vectors

prayush replied:
Just added the `inverse_spatial_metric_tag` to System, thanks!

nikwit commented on tests/Unit/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/Test_UpwindPenalty.cpp:
```diff
@@ -0,0 +1,310 @@
+  CurvedScalarWave::BoundaryCorrections::UpwindPenalty<Dim> csw_flux_computer{};
+
+  // buffer storing variables for CSW
+  TempBuffer<tmpl::list<
```
I see no reason to have Tempbuffers in a unit test. They reduce readability and we do not care about allocations/performance, so I would argue to remove it.

prayush replied:
Hmm. I sort of disagree on this. I would prefer leaving it here as is, if you don't mind too much. 

The reduction to readability of the test is limited to the declaration of the buffer itself, after which we are simply extracting references to memory in it - which just replaces allocating memory. 

And as spectre adds more and more unit tests, why would we not care about the performance of unit tests?

nikwit replied:
I think since the build time of unit tests is far longer than the runtime and this allocation is not in any kind of loop, the reduced memory allocation speed-up will not be noticeable at all.

Overall I just feel strongly that unit tests should be as explicit and clear as possible making them easy to debug and useful for understanding how the tested code works. But I dont want to impose my opinions here too strongly so I would be fine with approving it like this.

prayush replied:
That is fair. what also changed my mind is that this would be the only unit test that actually uses temp buffers, so I've removed their usage here as well.

nikwit reviewed (changes requested):
I still find a discrepancy between what I implemented and this one. I will try and hunt down the difference this weekend, but here are some comments for now.

Update: I have found the discrepancy to be a bug in my comparison code. This implementation and the one I did, which is a copy of the `ScalarWave` system with only characteristic speed calculation changed, produce the same results for random values even in curved space.

prayush committed:
Reduce 1 mem allocation in CSW::char_fields method

prayush commented:
Hi @nikwit @nilsdeppe : all your comments and suggestions have been incorporated / addressed. This PR is ready for another look.

nikwit commented:
You have a python-related CI failure in gcc8 debug which is the one that uses python2, so there is probably some compatibility issue there.

prayush committed:
Add overload of CSW::char_speeds that returns a tnsr<3> by ref

prayush commented:
Hi @nikwit : this is again ready for your perusal. @nilsdeppe already said above to squash all fixups directly. Let me know if you're okay with me squashing them now.

nikwit reviewed (dismissed):
This looks good to me now, please go ahead and squash

prayush committed:
Update CSW::System struct:

 - Add normal_dot_fluxes alias
 - Add inverse_metric tag to ensure correct curved-space
   normalization
 - Add alias to get largest char speed

prayush committed:
Add Upwind penalty BoundaryCorrection for CurvedScalarWave

prayush commented:
@nikwit - just squashed all fixups.

nikwit reviewed (approved):
looks good. Out of curiosity (and because I know this will happen to me at some point), what was the reason for the python 2 failure?

prayush commented:
> looks good. Out of curiosity (and because I know this will happen to me at some point), what was the reason for the python 2 failure?

Missing `__init__.py` files in the unit test directory!

prayush commented:
Hi @sxs-collaboration/spectre-core-devs - this PR is ready for a look / merge.
